### PR TITLE
implement native multisend

### LINKED ISSUE #423: [MultiSend] Implement native version
Need a new version of the MultiSend contract, `MultiSendNative`, that uses native tokens instead of ERC20

### PR DESCRIPTION
**Pull request by VargaElod23**

### CODE REVIEW

File: contracts/MultisendNative.sol [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            } else {
+                if (msg.value > total) {
+                    payable(msg.sender).transfer(msg.value - total);
+                    emit SentBack(msg.value - total);
+                }
+                break;
+            }
+        }
+        emit TokensSent(i, total);
+        return (i, total);
```
Comment (rattrap) [inline contracts/MultisendNative.sol]: Don't need to return anything

File: contracts/MultisendNative.sol [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        for (i; i < _recipients.length; ++i) {
+            if ((msg.value - total) >= _amounts[i]) {
+                _recipients[i].transfer(_amounts[i]);
+                total = total + _amounts[i];
+            } else {
+                if (msg.value > total) {
+                    payable(msg.sender).transfer(msg.value - total);
+                    emit SentBack(msg.value - total);
+                }
+                break;
```
Comment (rattrap) [inline contracts/MultisendNative.sol]: Too complicated and doesn't cover the case where we send this function more tokens than in the amounts array.

Let's just break here and move the send back functionality after the for

File: test/MultisendNative.test.js [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    assert.exists(event, 'TokensSent event not found');
+    event.args.total.toString().should.equal(totalWithoutLast.toString(), 'Total value mismatch');
+    event.args.lastSentIndex.toString().should.equal('9', 'Total value mismatch');
+
+    const eventSentBack = tx.logs.find((log) => log.event === 'SentBack');
+    assert.exists(eventSentBack, 'SentBack event not found');
+    eventSentBack.args.amount
+      .toString()
+      .should.equal(difference.toString(), 'Sent back value mismatch');
+  });
```
Comment (rattrap) [inline test/MultisendNative.test.js]: I think we can merge this with the previous test

**Review by rattrap — CHANGES_REQUESTED** [review]

File: test/MultisendNative.test.js (hunk 3 above)
Comment (VargaElod23) [reply test/MultisendNative.test.js]: I would like to keep these two as separate scenarios

**Comment by VargaElod23** [comment]
- [x] Removed fallback receiver

File: contracts/MultisendNative.sol [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+            'Multisend: contributors and balances have different sizes'
+        );
+        require(msg.value >= _amounts[0], "Multisend: Sent value doesn't cover first transaction");
+
+        uint256 total = 0;
+        uint8 i = 0;
+        for (i; i < _recipients.length; ++i) {
+            if ((msg.value - total) >= _amounts[i]) {
+                _recipients[i].transfer(_amounts[i]);
+                total = total + _amounts[i];
```
Comment (rattrap) [inline contracts/MultisendNative.sol]: Let's also add an event that contains address and amount

File: contracts/MultisendNative.sol [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            } else {
+                break;
+            }
+        }
+        if (msg.value > total) {
+            payable(msg.sender).transfer(msg.value - total);
+            emit SentBack(msg.value - total);
+        }
+        emit TokensSent(i, total);
```
Comment (rattrap) [inline contracts/MultisendNative.sol]: We can just remove this

File: contracts/MultisendNative.sol [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        require(
+            _recipients.length == _amounts.length,
+            'Multisend: contributors and balances have different sizes'
+        );
+        require(msg.value >= _amounts[0], "Multisend: Sent value doesn't cover first transaction");
+
+        uint256 total = 0;
+        uint8 i = 0;
+        for (i; i < _recipients.length; ++i) {
+            if ((msg.value - total) >= _amounts[i]) {
```
Comment (rattrap) [inline contracts/MultisendNative.sol]: Maybe we can reorganize this a bit - https://github.com/Alonski/MultiSendEthereum/blob/master/contracts/MultiSend.sol#L89

**Review by rattrap — CHANGES_REQUESTED** [review]

**Review by rattrap — APPROVED** [review]

**Pull request merged**